src/components/date/date.test.jsx

```diff
@@ -0,0 +1,376 @@
+
+    render(<Date {...requiredProps} />);
+    userEvent.type(screen.getByLabelText('Test label'), '1999-01-01');
+    expect(await screen.findByText(date.calendarHeading)).toBeInTheDocument();
+  });
+
+  describe('calendar behavior', () => {
+    it('toggles open/close', () => {
+      const today = getLocaleDate(new window.Date());
+
+      render(<Date {...requiredProps} />);
+      userEvent.click(screen.getByTitle('Test label calendar button'));
+      expect(screen.getByText(today.calendarHeading)).toBeInTheDocument();
+      userEvent.click(screen.getByTitle('Test label calendar button'));
+      expect(screen.queryByText(today.calendarHeading)).not.toBeInTheDocument();
+    });
+
+    it('closes on selection', () => {
+      const today = getLocaleDate(new window.Date());
+
+      render(<Date {...requiredProps} />);
+      userEvent.click(screen.getByTitle('Test label calendar button'));
+      userEvent.click(screen.getByLabelText(today.calendarDate));
+      expect(screen.queryByText(today.calendarDate)).not.toBeInTheDocument();
+      expect(screen.getByLabelText('Test label')).toHaveValue(today.displayValue);
+    });
+
+    it('closes on blur', () => {
+      const today = getLocaleDate(new window.Date());
+
+      render(<Date {...requiredProps} />);
+      userEvent.click(screen.getByTitle('Test label calendar button'));
+      expect(screen.getByText(today.calendarHeading)).toBeInTheDocument();
+      userEvent.click(document.body);
+      expect(screen.queryByText(today.calendarHeading)).not.toBeInTheDocument();
+      expect(document.activeElement).not.toBe(screen.getByLabelText('Test label'));
+    });
+
+    it('closes on escape', () => {
+      const today = getLocaleDate(new window.Date());
+
+      render(<Date {...requiredProps} />);
+      userEvent.type(screen.getByLabelText('Test label'), '');
+      expect(screen.getByText(today.calendarHeading)).toBeInTheDocument();
+      userEvent.type(screen.getByLabelText('Test label'), '{esc}', { skipClick: true });
+      expect(screen.queryByText(today.calendarHeading)).not.toBeInTheDocument();
+    });
+
+    it('does not open the native calendar picker', () => {
+      const onClickSpy = jest.fn(event => event.persist());
+      const onKeyDownSpy = jest.fn(event => event.persist());
+
+      render((
+        <div onClick={onClickSpy} onKeyDown={onKeyDownSpy}>
+          <Date {...requiredProps} />
+        </div>
+      ));
+
+      userEvent.click(screen.getByLabelText('Test label'));
+      userEvent.type(screen.getByLabelText('Test label'), ' ');
+
+      expect(onClickSpy).toBeCalledWith(expect.objectContaining({
+        defaultPrevented: true,
+        target: expect.anything(),
+      }));
+      expect(onKeyDownSpy).toBeCalledWith(expect.objectContaining({
+        defaultPrevented: true,
+        target: expect.anything(),
+      }));
+    });
+
+    it('updates the value between input types', () => {
```

Comment:
I don't quite follow what behavior we're testing here.

Reply:
It was an edge case where React would toggle between inputs but not update the value:

- render an `<input type="date" />` with value "2020-04-20"
- switch to `<input type="text" />` with value "2020-04-20" instead of "Apr 20, 2020"

The fix was specifying a `key` between input states. Good ol' React virtual DOM diffing. 

Does that make sense? Do you think I should change something to make it more apparent in the future?

Reply:
Oh wow, that's rough! Thanks React! 😆  It makes sense now, and now that I understand it the test name seems perfect... so I'm not really sure what to change it to. 🙃  Maybe adding a comment about how React wasn't switching the value when `type` changed for extra clarity?